### PR TITLE
[webgpu] Modularize binary and unary ops

### PR DESCRIPTION
**Pull request by haoyunfeix**
To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4465)

### CODE REVIEW

**Comment by haoyunfeix** [comment]
Mark status here:
Changes in this patch is for ops suitable for 2.7.0 and latest tfjs. For other ops, I will do in another PR when we change tfjs-core and tfjs-backend-cpu dependency from 2.7.0 to latest.

Binary completed: add, floorDiv, greaterEqual, less(newest support cpu shared), lessEqual, multiply, prelu, squaredDifference,
Binary half completed: sub(softmax depends on)
Binary uncompleted: greater, maximum (2.7.0 cpu not shared), realDiv(name changed after 2.7.0)

Unary completed: abs, log, relu, relu6, sigmoid, tanh
Unary half completed: exp (softmax depends on)
Unary uncompleted: neg (2.7.0 has no cpu shared)

**Comment by haoyunfeix** [comment]
@qjia7 @lina128 @kainino0x PTAL, thanks

File: tfjs-backend-webgpu/src/kernels/Div_impl.ts [hunk 1]
```diff
@@ -15,15 +15,14 @@
  * =============================================================================
  */
 
-import {Tensor, TensorInfo} from '@tensorflow/tfjs-core';
+import {TensorInfo} from '@tensorflow/tfjs-core';
 
 import {WebGPUBackend} from '../backend_webgpu';
 
-import {BinaryOpType, getBinaryProgram} from './binary_ops';
+import {BinaryOpType, getBinaryProgram} from '../kernels/binary_ops';
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/Div_impl.ts:22]: nit: Why can't directly use `./binary_ops`?

File: tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts [hunk 2]
```diff
@@ -0,0 +1,99 @@
+/**
+ * @license
+ * Copyright 2020 Google LLC. All Rights Reserved.
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts:3]: `2020` -> `2021`. And change it in all places. Thanks.

**Review by qjia7 — COMMENTED** [review]
LGTM with nits.
Would like @lina128 take a look. Thanks.

File: tfjs-backend-webgpu/src/kernels/Div_impl.ts (hunk 1 above)
Comment (haoyunfeix) [reply tfjs-backend-webgpu/src/kernels/Div_impl.ts]: Thanks for finding, fixed.

File: tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts (hunk 2 above)
Comment (haoyunfeix) [reply tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts]: fixed

File: tfjs-backend-webgpu/src/kernels/Abs.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+import {ABS} from './unary_op_webgpu';
+import {simpleAbsImplCPU} from '../kernel_utils/shared';
+
+export const abs =
+    unaryKernelFunc({opSnippet: ABS, cpuKernelImpl:simpleAbsImplCPU});
+
+export const absConfig: KernelConfig = {
+  kernelName: Abs,
+  backendName: 'webgpu',
+  kernelFunc: abs as {} as KernelFunc
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/kernels/Abs.ts:29]: nit: These casts seem like they shouldn't be necessary? But if they are, no problem.

**Review by kainino0x — COMMENTED** [review]
I don't have the context needed to review this, but LGTM otherwise. No particular comments.

**Review by lina128 — APPROVED** [review]
LGTM over all, just left some nit comments.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4465#-:-MRXbyp-4htvG4ajqE67:b5u21i4)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @haoyunfeix, @kainino0x, and @qjia7)

___
*[tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts, line 39 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXa1E54URfUAGBICzu:-MRXa1E54URfUAGBICzv:bk9chif) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts#L39)):*
> ```XML
>  */
> export function unaryKernelFunc(
>     {opSnippet, cpuKernelImpl, dtype}: UnaryKernelFuncConfig):
> ```

Can you add cpuKernelImpl description to doccomment?
___
*[tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts, line 53 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXaDSqCGhUKgffgjaT:-MRXaDSqCGhUKgffgjaU:bidglt0) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts#L53)):*
> ```XML
>     }
> 
>     const program: UnaryOpProgram = new UnaryOpProgram(x.shape, opSnippet);
> ```

Will you have UnaryOpPackedProgram in the future? If so, can add a TODO. Otherwise, please remove the param description from doccomment.
___
*[tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts, line 60 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXb0h84yLGw60oRMW9:-MRXb0h84yLGw60oRMWA:bsnnbll) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts#L60)):*
> ```XML
> type BinaryKernelFuncConfig = {
>   opSnippet: number,
>   boolType?: boolean,
> ```

Why do you need this? Can't it be derived from dtype?
___
*[tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts, line 69 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXadLB1rbSS6H4-tPA:-MRXadLC9o6LoD_S_Dk5:b-gfp4xv) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts#L69)):*
> ```XML
>  * @param opSnippet Op snippet to create `BinaryOpProgram`.
>  * @param packedOpSnippet Op snippet to create `BinaryOpPackedProgram`.
>  * @param checkOutOfBoundsForPackedProgram Whether to set checkOutOfBounds=true
> ```

Please update doccomment. There's a mismatch between the type declaration and doccomment.
___
*[tfjs-backend-webgpu/src/kernels/Abs.ts, line 29 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MPlnppVmyFf-tJdLECs-r4-29:-MRXbAJB01iyNVyrLTzc:bqtkv3v) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernels/Abs.ts#L29)):*
<details><summary><i>Previously, kainino0x (Kai Ninomiya) wrote…</i></summary><blockquote>

nit: These casts seem like they shouldn't be necessary? But if they are, no problem.
</blockquote></details>

KernelFunc has a more generic type. Each kernel (e.g. abs) uses more specific type, so that when kernels call each other internally, we have better type checking.
___
*[tfjs-backend-webgpu/src/kernels/GreaterEqual.ts, line 23 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXbgtn-ktjSE0GWqZr:-MRXbgtn-ktjSE0GWqZs:b-ygxpcf) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernels/GreaterEqual.ts#L23)):*
> ```XML
> 
> export const greaterEqual = binaryKernelFunc({
>   boolType: true,
> ```

Here, you can just pass `dtype: 'bool'`
___
*[tfjs-backend-webgpu/src/kernels/Less.ts, line 23 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXblSH9KPIqgfRNGpM:-MRXblSH9KPIqgfRNGpN:benpsw9) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernels/Less.ts#L23)):*
> ```XML
> 
> export const less = binaryKernelFunc({
>   boolType: true,
> ```

Same comment as above.

**Review by haoyunfeix — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4465)* status: :shipit: complete! 1 of 1 approvals obtained (waiting on @kainino0x, @lina128, and @qjia7)

___
*[tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts, line 3 at r1](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MPlnppVmyFf-tJdLECq-r1-3:-MRYR7JE9hLPejeYdriD:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/c629b6271239bee1eb7fb670e2703cb453548185/tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts#L3)):*
<details><summary><i>Previously, haoyunfeix (Hao Yunfei) wrote…</i></summary><blockquote>

fixed
</blockquote></details>

Done.
___
*[tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts, line 39 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXa1E54URfUAGBICzu:-MRYQe62BiCF22mzBVZP:bt2vr1b) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts#L39)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Can you add cpuKernelImpl description to doccomment?
</blockquote></details>

Fixed.
___
*[tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts, line 53 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXaDSqCGhUKgffgjaT:-MRYQidq8aN8nCw3pOOK:bss4a2b) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts#L53)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Will you have UnaryOpPackedProgram in the future? If so, can add a TODO. Otherwise, please remove the param description from doccomment.
</blockquote></details>

Removed.
___
*[tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts, line 60 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXb0h84yLGw60oRMW9:-MRYQmAO3f7lTvgl8_0N:bnr3vsn) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts#L60)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Why do you need this? Can't it be derived from dtype?
</blockquote></details>

We don't need this, thanks for the finding. Fixed.
___
*[tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts, line 69 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXadLB1rbSS6H4-tPA:-MRYQvykEEvJdY8Ah-4V:b-bw8e14) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernel_utils/kernel_funcs_utils.ts#L69)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Please update doccomment. There's a mismatch between the type declaration and doccomment.
</blockquote></details>

Updated.
___
*[tfjs-backend-webgpu/src/kernels/GreaterEqual.ts, line 23 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXbgtn-ktjSE0GWqZr:-MRYRA7395VcCEpJnAmI:b-n4kr7n) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernels/GreaterEqual.ts#L23)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Here, you can just pass `dtype: 'bool'`
</blockquote></details>

Updated. Thanks.
___
*[tfjs-backend-webgpu/src/kernels/Less.ts, line 23 at r4](https://reviewable.io/reviews/tensorflow/tfjs/4465#-MRXblSH9KPIqgfRNGpM:-MRYRF2j4CXphO4UgReX:b-896fix) ([raw file](https://github.com/tensorflow/tfjs/blob/0b71733e07e11ae870a196092947a70eda53c170/tfjs-backend-webgpu/src/kernels/Less.ts#L23)):*
<details><summary><i>Previously, lina128 (Na Li) wrote…</i></summary><blockquote>

Same comment as above.
</blockquote></details>

Done.

**Review by qjia7 — APPROVED** [review]
Thanks for the final cleanup. LGTM. Will land this PR in case not blocking your next modularization work.
